### ISSUE TITLE
podman pull randomly produces panic runtime error

### ISSUE THREAD

**Issue by EmilienM** [containers/libpod#1730]
**Is this a BUG REPORT or FEATURE REQUEST?**:

[//]: # **kind bug**

**Description**
In OpenStack CI, we replaced Docker by Podman, and we now see random runtime error when running ```podman pull```.

**Steps to reproduce the issue:**
It is hard to reproduce but we see it on a node that is short on memory, pulling large images.

**Describe the results you received:**

```
    "panic: runtime error: invalid memory address or nil pointer dereference",
    "[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xc46e67]",
    "goroutine 1 [running]:",
    "github.com/containers/libpod/vendor/github.com/containers/image/storage.(*storageImageDestination).Commit(0xc42045c1c0, 0x13eee40, 0xc420034140, 0x0, 0x0)",
    "\t/builddir/build/BUILD/libpod-6e1aeb06f86bfed7045be19c8e8b09c1bf5ba55f/_build/src/github.com/containers/libpod/vendor/github.com/containers/image/storage/storage_image.go:601 +0x407",
    "github.com/containers/libpod/vendor/github.com/containers/image/copy.Image(0x13eee40, 0xc420034140, 0xc42042a940, 0x13fa040, 0xc42013c780, 0x13fa340, 0xc4204201a0, 0xc42013c9c0, 0xc4200fc000, 0x1dca, ...)",
    "\t/builddir/build/BUILD/libpod-6e1aeb06f86bfed7045be19c8e8b09c1bf5ba55f/_build/src/github.com/containers/libpod/vendor/github.com/containers/image/copy/copy.go:178 +0x59e",
    "github.com/containers/libpod/libpod/image.(*Runtime).doPullImage(0xc4201d0980, 0x13eee40, 0xc420034140, 0xc420330500, 0xc420258810, 0x1, 0x1, 0x0, 0x0, 0x0, ...)",
    "\t/builddir/build/BUILD/libpod-6e1aeb06f86bfed7045be19c8e8b09c1bf5ba55f/_build/src/github.com/containers/libpod/libpod/image/pull.go:258 +0x333",
    "github.com/containers/libpod/libpod/image.(*Runtime).pullImageFromHeuristicSource(0xc4201d0980, 0x13eee40, 0xc420034140, 0x7ffed5b927e4, 0x5f, 0x13dd4a0, 0xc42000e020, 0x0, 0x0, 0x0, ...)",
    "\t/builddir/build/BUILD/libpod-6e1aeb06f86bfed7045be19c8e8b09c1bf5ba55f/_build/src/github.com/containers/libpod/libpod/image/pull.go:212 +0x1e0",
    "github.com/containers/libpod/libpod/image.(*Runtime).New(0xc4201d0980, 0x13eee40, 0xc420034140, 0x7ffed5b927e4, 0x5f, 0x0, 0x0, 0x0, 0x0, 0x13dd4a0, ...)",
    "\t/builddir/build/BUILD/libpod-6e1aeb06f86bfed7045be19c8e8b09c1bf5ba55f/_build/src/github.com/containers/libpod/libpod/image/image.go:148 +0x1c1",
    "main.pullCmd(0xc4201266e0, 0x0, 0x0)",
    "\t/builddir/build/BUILD/libpod-6e1aeb06f86bfed7045be19c8e8b09c1bf5ba55f/_build/src/github.com/containers/libpod/cmd/podman/pull.go:128 +0x825",
    "github.com/containers/libpod/vendor/github.com/urfave/cli.HandleAction(0x10efc00, 0x13381d0, 0xc4201266e0, 0x0, 0xc4202ba480)",
    "\t/builddir/build/BUILD/libpod-6e1aeb06f86bfed7045be19c8e8b09c1bf5ba55f/_build/src/github.com/containers/libpod/vendor/github.com/urfave/cli/app.go:501 +0xc8",
    "github.com/containers/libpod/vendor/github.com/urfave/cli.Command.Run(0x12a8812, 0x4, 0x0, 0x0, 0x0, 0x0, 0x0, 0x12cb595, 0x1d, 0x0, ...)",
    "\t/builddir/build/BUILD/libpod-6e1aeb06f86bfed7045be19c8e8b09c1bf5ba55f/_build/src/github.com/containers/libpod/vendor/github.com/urfave/cli/command.go:165 +0x47d",
    "github.com/containers/libpod/vendor/github.com/urfave/cli.(*App).Run(0xc42018ea80, 0xc420030090, 0x3, 0x3, 0x0, 0x0)",
    "\t/builddir/build/BUILD/libpod-6e1aeb06f86bfed7045be19c8e8b09c1bf5ba55f/_build/src/github.com/containers/libpod/vendor/github.com/urfave/cli/app.go:259 +0x6e8",
    "main.main()",
    "\t/builddir/build/BUILD/libpod-6e1aeb06f86bfed7045be19c8e8b09c1bf5ba55f/_build/src/github.com/containers/libpod/cmd/podman/main.go:263 +0x1816",
    "2018-10-30 20:06:20,502 WARNING: 9470 -- retrying pulling image: 192.168.24.1:8787/tripleomaster/centos-binary-ironic-api:current-tripleo-updated-20181030192232",
    "2018-10-30 20:06:20,874 WARNING: 9469 -- podman pull failed: Trying to pull 192.168.24.1:8787/tripleomaster/centos-binary-heat-api:current-tripleo-updated-20181030192232...Getting image source signatures",
```

**Describe the results you expected:**

Podman pull should work.

**Additional information you deem important (e.g. issue happens only occasionally):**
dstat output:
http://logs.openstack.org/58/609858/6/gate/tripleo-ci-centos-7-containers-multinode/27bedd3/logs/undercloud/var/log/extra/dstat.html.gz

See at 20:06 when the pull fails.

**Output of `podman version`:**

```
podman-0.10.1.3-2.git6e1aeb0.el7.x86_64
```

**Output of `podman info`:**
```
host:
  BuildahVersion: 1.5-dev
  Conmon:
    package: podman-0.10.1.3-2.git6e1aeb0.el7.x86_64
    path: /usr/libexec/podman/conmon
    version: 'conmon version 1.12.0, commit: 097cf71375bd18454d1c48b9fdf8ccff2ed995f8-dirty'
  Distribution:
    distribution: '"centos"'
    version: "7"
  MemFree: 576225280
  MemTotal: 8365137920
  OCIRuntime:
    package: runc-1.0.0-54.dev.git2abd837.el7.x86_64
    path: /usr/bin/runc
    version: 'runc version spec: 1.0.0'
  SwapFree: 8560570368
  SwapTotal: 8588881920
  arch: amd64
  cpus: 8
  hostname: undercloud.localdomain
  kernel: 3.10.0-862.14.4.el7.x86_64
  os: linux
  uptime: 1h 5m 46.27s (Approximately 0.04 days)
insecure registries:
  registries:
  - 192.168.24.1:8787
  - 192.168.24.3:8787
registries:
  registries:
  - docker.io
  - registry.fedoraproject.org
  - quay.io
  - registry.access.redhat.com
  - registry.centos.org
store:
  ContainerStore:
    number: 1
  GraphDriverName: overlay
  GraphOptions:
  - overlay.override_kernel_check=true
  GraphRoot: /var/lib/containers/storage
  GraphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
  ImageStore:
    number: 19
  RunRoot: /var/run/containers/storage
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Logs of the failure:
http://logs.openstack.org/58/609858/6/gate/tripleo-ci-centos-7-containers-multinode/27bedd3/logs/undercloud/home/zuul/install-undercloud.log.txt.gz

**Comment by mheon** [comment]
@mtrmac Looking at this one, it looks like it's happening in c/image code

Per discussion in IRC, the node is under memory pressure - possibly an OOM somewhere?

**Comment by sshnaidm** [comment]
@mheon  I don't see any OOM in journal log: http://logs.openstack.org/58/609858/6/gate/tripleo-ci-centos-7-containers-multinode/27bedd3/logs/undercloud/var/log/journal.txt.gz

**Comment by mtrmac** [comment]
AFAICS this is https://github.com/containers/image/blob/4b6d9cf918c1e79a605265568c93cb35185ad25d/storage/storage_image.go#L597-L601 : if a layer already exists locally, `PutLayer` returns `(nil, _, storage.ErrDuplicateID)`.

@nalind Should the code just call `store.Layer` to look up the layer by the ID in that case?

**Comment by runcom** [comment]
@mtrmac fix should be https://github.com/containers/storage/pull/234 right?

**Comment by mtrmac** [comment]
> @mtrmac fix should be containers/storage#234 right?

Sure, that looks even more efficient.

**Comment by EmilienM** [comment]
@mheon let me know if I'm doing wrong but I think we can close this one since https://github.com/containers/storage/pull/234 seems to be the fix. If yes, feel free to close. Thanks
